### DISCUSSION TITLE
How to create databases?

### DISCUSSION THREAD

**Discussion by bmulholland** [ioquatix/turbo_test#8] [Q&A]
I tried installing and running with the steps in https://github.com/ioquatix/turbo_test/issues/6, but all the parallel databases do not exist, so the run fails.

`rake parallel:create` does not exist, and the source doesn't seem to mention another way of creating the dbs. Once created, they'll need to be updated with migrations etc too - but that doesn't seem to be discussed anywhere?

**Comment by ioquatix** [comment]
There are many ways to do this, so it depends on your circumstance and how your tests work.

Here is an example:

``` ruby
ENV['RAILS_ENV'] = 'test'

worker do |index|
  ENV['TEST_ENV_NUMBER'] = index.to_s

  system("bin/rails", "db:drop", "db:create", "db:migrate")

  # Parallel spec system
  port = 10000 + index
  
  Console.logger.info "Setting up parallel test mode - starting Redis #{n} on port #{port}"
  
  system("rm -rf tmp/test_data_#{n} && mkdir -p tmp/test_data_#{n}/redis")
  pid = Process.spawn("redis-server --dir tmp/test_data_#{n}/redis --port #{port}", out: "/dev/null")
  
  ENV["REDIS_PORT"] = port.to_s
  ENV["RAILS_DB"] = "discourse_test_#{n}"
    
  at_exit do
    Process.kill("SIGTERM", pid)
    Process.wait
  end
  
  require_relative 'spec/rails_helper'
end
```

**Reply by bmulholland** [reply]
What about the tests would change the approach? In which circumstance would I want to do something differently? How would the setup steps be different?

**Reply by ioquatix** [reply]
Databases represent persistent state.

Depending on your tests, you may need to clear that state after each run... or not.

I tend to write my tests so that the current state of the database doesn't matter.

This avoids the need for complex transaction-per-test or "database cleaner" gems.

If you do this, simply pointing all your tests at a single database instance can work. So you don't need to create it every time, or multiple separate instances for different test runners.

Some test suites DO need a separate database per test running in parallel. If that's the case, you'll either need to create it before running the tests and reuse it, or create it as part of the test script. Personally, I like it to be an atomic unit, so I'd incorporate it into the worker setup.

In the end, it's up to you to decide what makes the most sense.

**Reply by bmulholland** [reply]
Thanks, I updated https://github.com/ioquatix/turbo_test/issues/6 with the bare-bones version of that.
